### PR TITLE
feat: add event-driven chaos triggers with command trigger (Phase 1)

### LINKED ISSUE #1483: feat: event-driven chaos triggers — wait for conditions before injection
## Summary

Add an optional `triggers` mechanism that lets users define preconditions (Kubernetes resource state, Prometheus metrics, shell commands, HTTP endpoints) that must be satisfied before krkn begins chaos injection. Today krkn fires immediately after initialization — this feature adds a "wait for ready" gate at the krkn run level.

## Motivation

- Deployment rollouts, canary promotions, and scaling events take time — chaos should start only after the target is in the expected state
- Load-driven chaos (inject faults when CPU > 80%) requires watching metrics before triggering
- Current workaround is wrapping krkn in external scripts, which is brittle and not observable
- **KubeVirt live migration chaos**: During cross-cluster live migration (CCLM), the actual migration window is very short (30-40 seconds). Today, krkn's ~8 second warmup plus timing guesswork makes it nearly impossible to inject chaos precisely during that window. With event-driven triggers, krkn can watch the VirtualMachineInstanceMigration resource and inject chaos the moment migration enters the `Running` phase — hitting the exact window that matters

### Trigger Types

| Type | What it checks | Example use case |
|------|---------------|-----------------|
| **`command`** | Shell command exit code | Run any custom readiness check script |
| **`k8s`** | Kubernetes resource state via JSONPath | Wait for Deployment replicas >= 3, VMI migration phase == Running |
| **`prometheus`** | PromQL query result | Trigger when CPU utilization crosses threshold |
| **`http`** | HTTP endpoint status code | Wait for app health endpoint to return 200 |

### Config Example

Triggers are defined as a top-level block in `config.yaml`, gating the entire chaos run. Once all conditions are met, scenarios execute. No `triggers` block = current behavior (fully backward compatible).

```yaml
triggers:
  mode: all_of        # all_of (default) | any_of
  timeout: 300        # max wait in seconds (default 300)
  interval: 5         # polling interval in seconds (default 5)
  on_timeout: skip    # skip | fail | run_anyway
  conditions:
    - type: command
      inline: "kubectl get deploy nginx -n default -o jsonpath='{.status.readyReplicas}' | grep -q 3"

chaos_scenarios:
  - pod_disruption_scenarios:
      - scenarios/kube/pod.yml
```

### Trigger Options

- **`mode`**: `all_of` (all conditions must pass) or `any_of` (any one suffices)
- **`timeout`**: Maximum seconds to wait before giving up
- **`interval`**: Polling frequency in seconds
- **`on_timeout`**: What happens when timeout is reached — `skip` the scenario, `fail` it, or `run_anyway`

### Architecture

- **Pluggable**: `AbstractTrigger` base class — new trigger types just extend it and implement `evaluate()` + `describe()`
- **Polling-based** — simpler, works uniformly across all trigger types
- Trigger evaluation happens in `run_kraken.py` before the chaos loop begins
- Extends the existing signal server (`/status` endpoint) to report trigger state

### krkn-hub Integration

- New `TRIGGER_*` environment variables (`TRIGGER_ENABLED`, `TRIGGER_MODE`, `TRIGGER_CMD_INLINE`, etc.)
- Flows through existing `env.sh` → `envsubst` → YAML template pattern
- Since krkn-hub runs one scenario per container, a top-level trigger naturally maps to the execution model

### krknctl Integration

- New `--trigger-*` CLI flags mapped to `TRIGGER_*` env vars
- Trigger fields in `krknctl-input.json` for the interactive form (`--form`)
- `krknctl query-status` extended to show trigger waiting state

### Phased Delivery

1. **Phase 1 (MVP):** `command` trigger + `TriggerManager` + core integration
2. **Phase 2:** `k8s` trigger
3. **Phase 3:** `prometheus` trigger
4. **Phase 4:** `http` trigger
5. **Phase 5:** Signal server status extension + krknctl status reporting

### PR DESCRIPTION
**Pull request by ddjain**
## Summary

Adds a pluggable trigger system that gates chaos injection on user-defined preconditions. A top-level `triggers` block in `config.yaml` is evaluated before the chaos loop — krkn polls conditions and only begins scenario execution once they're satisfied. No `triggers` block = current behavior (fully backward compatible).

Phase 1 implements the **command** trigger type, which runs a shell command and checks its exit code.

## Config Example

```yaml
triggers:
  mode: all_of        # all_of (default) | any_of
  timeout: 300        # max wait in seconds (default 300)
  interval: 5         # polling interval in seconds (default 5)
  on_timeout: skip    # skip | fail | run_anyway
  conditions:
    - type: command
      inline: "kubectl get deploy nginx -o jsonpath='{.status.readyReplicas}' | grep -q 3"
```

## Changes

- **`krkn/scenario_plugins/triggers/`** — New trigger module:
  - `AbstractTrigger` base class (pluggable — future types just extend it)
  - `CommandTrigger` — runs shell command, checks exit code, logs on state change
  - `TriggerManager` — orchestrates polling with mode/timeout/interval/on_timeout
  - Input validation for timeout/interval (positive numbers) and expected_rc (int coercion)
- **`run_kraken.py`** — Top-level trigger evaluation before the chaos loop; uses `return 1` (not `sys.exit`) for consistency with existing error paths
- **`tests/test_triggers/`** — 32 unit tests covering all trigger functionality and validation edge cases

## Related Tickets & Documents

- Closes #1483

## Documentation

- [x] **Is documentation needed for this update?**

## Checklist before requesting a review

- [x] I have performed a self-review of my code by running krkn and specific scenario
- [x] If it is a core feature, I have added thorough unit tests with above 80% coverage

### Test Results

**Unit tests (32 trigger tests, 1243 total):**
```
Ran 1243 tests in 16.111s
OK
```

**Live GCP cluster test** — deployed nginx with 1 replica, configured trigger to wait for 3 replicas, krkn waited ~13s then injected pod kill chaos on scale-up:
```
waiting for triggers: mode=all_of, timeout=120.0s, interval=5.0s, on_timeout=fail
trigger condition not satisfied: kubectl ... | grep -q 3
  (scaled nginx to 3 replicas)
trigger condition satisfied: kubectl ... | grep -q 3
all trigger conditions satisfied
Running PodDisruptionScenarioPlugin -> pod_scenario.yml
Deleting pod nginx-66686b6766-5h5d7
exit code: 0
```

### CODE REVIEW

File: run_kraken.py [hunk 1]
```diff
@@ -419,6 +420,36 @@ def main(options, command: Optional[str]) -> int:
...
+                    if on_timeout == "skip":
+                        logging.warning(
+                            "trigger timed out, skipping all scenarios"
+                        )
+                        chaos_scenarios = []
+                    elif on_timeout == "fail":
+                        logging.error(
+                            "trigger timed out, exiting with failure"
+                        )
+                        return 1
```
[the remark this answers is not in the document]
Comment (ddjain) [reply run_kraken.py:437]: Fixed — moved trigger evaluation before `health_check_factory.start_all()` so no worker threads are spawned when `return 1` fires.

File: krkn/scenario_plugins/triggers/command_trigger.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                text=True,
+            )
+            met = result.returncode == self._expected_rc
+        except subprocess.TimeoutExpired:
+            logging.warning(
+                f"command trigger timed out after {COMMAND_TIMEOUT_SECONDS}s: "
+                f"{self._cmd}"
+            )
+            met = False
+        except FileNotFoundError:
```
[the remark this answers is not in the document]
Comment (ddjain) [reply krkn/scenario_plugins/triggers/command_trigger.py:67]: Fixed — added a broad `except Exception` fallback that logs and sets `met = False`, preventing `UnboundLocalError` from crashing the polling loop.

File: krkn/scenario_plugins/triggers/command_trigger.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            )
+        self._last_result: bool | None = None
+
+    def evaluate(self) -> bool:
+        try:
+            result = subprocess.run(
+                self._cmd,
+                shell=True,
+                timeout=COMMAND_TIMEOUT_SECONDS,
+                capture_output=True,
```
[the remark this answers is not in the document]
Comment (ddjain) [reply krkn/scenario_plugins/triggers/command_trigger.py:49]: Fixed — added debug-level logging of `returncode` and `stderr` on every poll, making failing triggers diagnosable without enabling verbose mode.

File: krkn/scenario_plugins/triggers/trigger_manager.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            raise ValueError(
+                f"timeout must be positive, got {self._timeout}"
+            )
+        if self._interval <= 0:
+            raise ValueError(
+                f"interval must be positive, got {self._interval}"
+            )
+
+        self._triggers: list[AbstractTrigger] = []
+        for condition in trigger_config["conditions"]:
```
Comment (tsebastiani) [inline krkn/scenario_plugins/triggers/trigger_manager.py:81]: I would check if `trigger_config["conditions"]` is a list, if for any reason it is a string it will iterate over the characters of the string and will lead to unpredictable behaviour

File: run_kraken.py [hunk 5]
```diff
@@ -412,6 +413,36 @@ def main(options, command: Optional[str]) -> int:
...
+        if trigger_config:
+            try:
+                trigger_manager = TriggerManager(trigger_config)
+                logging.info(
+                    "waiting for triggers before starting chaos:\n%s",
+                    trigger_manager.describe(),
+                )
+                triggered = trigger_manager.wait_for_triggers()
+                if not triggered:
+                    on_timeout = trigger_config.get("on_timeout", "skip")
```
Comment (tsebastiani) [inline run_kraken.py]: Here you read `on_timeout` from the config skipping the `TriggerManager` validation mechanism, why don't you read from `trigger_manager._on_timeout` (or implement a `get_timeout()` method in the class to retrieve it?

File: krkn/scenario_plugins/triggers/command_trigger.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    """Trigger that runs a shell command and checks its exit code."""
+
+    def __init__(self, config: dict):
+        self._cmd = config.get("cmd") or config.get("inline")
+        if not self._cmd:
+            raise ValueError(
+                "command trigger requires either 'cmd' or 'inline' field"
+            )
+        try:
+            self._expected_rc = int(config.get("expected_rc", 0))
```
Comment (tsebastiani) [inline krkn/scenario_plugins/triggers/command_trigger.py:32]: here the validation only checks if expected_rc is an int but should be also checked if it is between 0 and 255 (unix return codes are only in this range) and the condition will be never met even if the rc passes the validation

File: run_kraken.py (hunk 5 above)
Comment (ddjain) [reply run_kraken.py]: added an on_timeout property on TriggerManager and updated run_kraken.py to read from trigger_manager.on_timeout instead of re-reading from raw config.

File: krkn/scenario_plugins/triggers/command_trigger.py (hunk 6 above)
Comment (ddjain) [reply krkn/scenario_plugins/triggers/command_trigger.py:32]: added range validation 0 <= expected_rc <= 255.

File: krkn/scenario_plugins/triggers/trigger_manager.py (hunk 4 above)
Comment (ddjain) [reply krkn/scenario_plugins/triggers/trigger_manager.py:81]: Added isinstance(conditions, list) check that raises ValueError

**Review by tsebastiani — APPROVED** [review]
/LGTM
